rest/oidc_api_test.go

```diff
@@ -863,3 +870,266 @@ func assertHttpResponse(t *testing.T, response *http.Response, statusCode int, b
+					IncludeAccessToken:   true,
+					DisableCallbackState: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+		}, {
+			name: "callback state disabled explicitly with HttpOnly cookie",
+			inputProviders: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:                        "foo",
+					Issuer:                      issuerFoo,
+					ClientID:                    "baz",
+					UserPrefix:                  "foo",
+					ValidationKey:               base.StringPointer("qux"),
+					Register:                    true,
+					DiscoveryURI:                issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken:          true,
+					DisableCallbackState:        true,
+					CallbackStateCookieHTTPOnly: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+		}, {
+			name: "mitigation from csrf when callback state enabled",
+			inputProviders: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+			invalidateState:      true,
+		}, {
+			name: "vulnerable to csrf when callback state disabled",
+			inputProviders: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("quz"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			inputDefaultProvider: "foo",
+			wantUsername:         "foo_noah",
+			inputRequestURL:      "/db/_oidc?provider=foo&offline=true",
+			invalidateState:      true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			opts := auth.OIDCOptions{Providers: tc.inputProviders, DefaultProvider: &tc.inputDefaultProvider}
+			restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &opts}}
+			restTester := NewRestTester(t, &restTesterConfig)
+			defer restTester.Close()
+			mockSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+			defer mockSyncGateway.Close()
+			mockSyncGatewayURL := mockSyncGateway.URL
+
+			if tc.invalidateState {
+				mockAuthServer.SetOption("invalidateState", true)
+				defer mockAuthServer.DeleteOption("invalidateState")
+			}
```

Comment:
Same comments here about the shared `mockAuthServer` between tests as on the CBG-804 PR, parameterise the constructor of mockAuthServer and create a new instance for each subtest to prevent shared state.

Reply:
This change has been revised in latest revision.